### PR TITLE
[C API] Initial interface proposal

### PR DESCRIPTION
**Pull request by VivekPanyam**
### Summary:

An initial interface for a Neuropod C API.

### CODE REVIEW

File: source/neuropod/bindings/c/np_valuemap.h [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+// Insert a tensor into the provided map
+// Overwrites any existing tensors with the same name
+// Note: this does not transfer ownership and the caller is still responsible for calling NP_DecrementRefcount on
+// `tensor`
+void NP_InsertTensor(NP_NeuropodValueMap *nvm, const char *name, NP_NeuropodTensor *tensor);
+
+// Given a NP_NeuropodValueMap, find a tensor with name `name` and return it
+// If such a tensor is not found, returns nullptr
+// Note: the caller is responsible for calling NP_DecrementRefcount on the returned tensor
+NP_NeuropodTensor *NP_GetTensor(const NP_NeuropodValueMap *nvm, const char *name);
```
Comment (VivekPanyam) [inline source/neuropod/bindings/c/np_valuemap.h:47]: Need to add a way to get the names of all tensors in a `NP_NeuropodValueMap`

File: source/neuropod/bindings/c/np_tensor.h [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+// Get the length of a specified element in a string tensor
+void NP_GetStringElementSize(const NP_NeuropodTensor *tensor, size_t index, size_t *elem_size, NP_Status *status);
+
+// Copy the value of a specified element of a string tensor into a user provided buffer
+void NP_GetStringElement(
+    const NP_NeuropodTensor *tensor, size_t index, char *buffer, size_t buffer_size, NP_Status *status);
+
+// Drops the internal refcount of the underlying tensor
+// This should be called on any tensor returned by the C API.
```
Comment (VivekPanyam) [inline source/neuropod/bindings/c/np_tensor.h:75]: Add something like

```
After a call to this function, `tensor` is no longer a pointer to a valid object
```

File: source/neuropod/bindings/c/c_api.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+// Note: The caller is responsible for freeing the returned TensorSpec
+NP_TensorSpec *NP_GetInputSpec(NP_Neuropod *model, size_t index);
+
+// Get an item from the output spec of a model
+// Returns nullptr if index is out of range
+// Note: The caller is responsible for freeing the returned TensorSpec
+NP_TensorSpec *NP_GetOutputSpec(NP_Neuropod *model, size_t index);
+
+// Get an allocator for a model
+NP_TensorAllocator *NP_GetAllocator(NP_Neuropod *model);
```
Comment (VivekPanyam) [inline source/neuropod/bindings/c/c_api.h:87]: Add
```
Note: The caller is responsible for freeing the returned TensorAllocator
```

**Comment by VivekPanyam** [comment]
Updated based on comments from @selitvin. Any thoughts @vkuzmin-uber?

File: source/neuropod/bindings/c/np_tensor_allocator.h [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+#pragma once
+
+#include "neuropod/bindings/c/np_tensor.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct NP_TensorAllocator NP_TensorAllocator;
```
Comment (vkuzmin-uber) [inline source/neuropod/bindings/c/np_tensor_allocator.h:26]: I don't follow this and other declarations above for NP_Status, NP_Neuropod, NP_NeuropodTensor- is it just a stub and actual declarations will be added later?

File: source/neuropod/bindings/c/c_api.h [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+void NP_Infer(NP_Neuropod *model, const NP_NeuropodValueMap *inputs, NP_NeuropodValueMap **outputs, NP_Status *status);
+
+// Run inference with a set of requested outputs
+// requested_outputs should be a null terminated array of char ptr containing the names of requested outputs
+// Note: The caller is responsible for freeing the returned NP_NeuropodValueMap
+void NP_InferWithRequestedOutputs(NP_Neuropod *              model,
+                                  const NP_NeuropodValueMap *inputs,
+                                  const char **              requested_outputs,
+                                  NP_NeuropodValueMap **     outputs,
+                                  NP_Status *                status);
```
Comment (vkuzmin-uber) [inline source/neuropod/bindings/c/c_api.h:64]: who is responsible for allocating and "free" "NP_Status * status" ?

File: source/neuropod/bindings/c/np_tensor_allocator.h (hunk 4 above)
Comment (VivekPanyam) [reply source/neuropod/bindings/c/np_tensor_allocator.h:26]: These are the public facing header files (so these structs are opaque to users).

The definitions will be in `_internal.h` files which will be added in another PR

See the TensorFlow C API for an [example](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/tf_status_internal.h) of this pattern

File: source/neuropod/bindings/c/c_api.h (hunk 5 above)
Comment (VivekPanyam) [reply source/neuropod/bindings/c/c_api.h:64]: The caller is responsible for doing so (using `NP_NewStatus` and `NP_DeleteStatus`)

**Pull request merged**